core/drivers/remoteproc/stm32_remoteproc.c

```diff
@@ -0,0 +1,390 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2023, STMicroelectronics
+ */
+
+#include <assert.h>
+#include <config.h>
+#include <drivers/rstctrl.h>
+#include <drivers/stm32_remoteproc.h>
+#include <kernel/cache_helpers.h>
+#include <kernel/dt_driver.h>
+#include <kernel/tee_misc.h>
+#include <libfdt.h>
+#include <mm/core_memprot.h>
+#include <mm/core_mmu.h>
+
+#define TIMEOUT_US_1MS	U(1000)
+
+/**
+ * struct stm32_rproc_mem - Memory regions used by the remote processor
+ *
+ * @addr:	physical base address from the CPU space perspective
+ * @da:		device address corresponding to the physical base address
+ *		from remote processor space perspective
+ * @size:	size of the region
+ */
+struct stm32_rproc_mem {
+	paddr_t addr;
+	paddr_t da;
+	size_t size;
+};
+
+/**
+ * struct stm32_rproc_instance - rproc instance context
+ *
+ * @cdata:	pointer to the device compatible data
+ * @link:	the node in the rproc_list
+ * @regions:	memory regions used
+ * @mcu_rst:	remote processor reset control
+ * @hold_boot:	remote processor hold boot control
+ * @n_regions:	number of memory regions
+ */
+struct stm32_rproc_instance {
+	const struct stm32_rproc_compat_data *cdata;
+	SLIST_ENTRY(stm32_rproc_instance) link;
+	struct stm32_rproc_mem *regions;
+	struct rstctrl *mcu_rst;
+	struct rstctrl *hold_boot;
+	size_t n_regions;
+};
+
+/**
+ * struct stm32_rproc_compat_data - rproc associated data for compatible list
+ *
+ * @firmware_id:	identifies the remoteproc firmware to load
+ */
+struct stm32_rproc_compat_data {
+	uint32_t firmware_id;
+};
+
+static SLIST_HEAD(, stm32_rproc_instance) rproc_list =
+		SLIST_HEAD_INITIALIZER(rproc_list);
+
+void *stm32_rproc_get(uint32_t fw_id)
+{
+	struct stm32_rproc_instance *rproc = NULL;
+
+	SLIST_FOREACH(rproc, &rproc_list, link)
+		if (rproc->cdata->firmware_id == fw_id)
+			return rproc;
+
+	return NULL;
+}
+
+TEE_Result stm32_rproc_start(uint32_t firmware_id)
+{
+	struct stm32_rproc_instance *rproc = stm32_rproc_get(firmware_id);
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	if (!rproc || !rproc->hold_boot)
+		return TEE_ERROR_GENERIC;
+
+	/*
+	 * The firmware is started by de-asserting the hold boot and
+	 * asserting it back to avoid auto restart on a crash.
+	 * No need to release the MCU reset as it is automatically released by
+	 * the hardware.
+	 */
+	res = rstctrl_deassert_to(rproc->hold_boot, TIMEOUT_US_1MS);
+	if (res)
+		return res;
+
+	res = rstctrl_assert_to(rproc->hold_boot, TIMEOUT_US_1MS);
+	if (res)
+		return res;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result __stm32_rproc_stop(struct stm32_rproc_instance *rproc)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	if (!rproc->hold_boot || !rproc->mcu_rst)
+		return TEE_ERROR_GENERIC;
+
+	res = rstctrl_assert_to(rproc->hold_boot, TIMEOUT_US_1MS);
+	if (res)
+		return res;
+
+	res = rstctrl_assert_to(rproc->mcu_rst, TIMEOUT_US_1MS);
+	if (res)
+		return res;
+
+	return TEE_SUCCESS;
```

Comment:
minor suggestion, to shorten the function:
```c
	res = rstctrl_assert_to(rproc->hold_boot, TIMEOUT_US_1MS);
	if (!res)
		res = rstctrl_assert_to(rproc->mcu_rst, TIMEOUT_US_1MS);

	return res;
```

Reply:
Done